Linking exp decay

aclark29 opened the pull request:
Summary:
- Added the functionality for the user to configure the model to use either "multiplier" or "exponential" modes of decay.
- Edited the sim.conf example to reflect this.
- Added 2 Linking tests, one for checking the probability of linking with exponential enabled, and one for checking the probability with multiplier enabled.
- Added a test to check that when time_last_active is set to -1 when inputs for time_last_active_drug_use are <= -1.

Old:
- A single multiplier was used to scale linking probability for those who had recently been screened. Within the recent cutoff window the rate will be L' = L * multiplier.

New:
- sim.conf contains a field named "decay_type" which takes "exponential" or "multiplier".
- If "exponential" is used, the rate decreases by the equation L' = L * e^(-t)
- If "multiplier" is used the original multiplier feature will be used.
- If left blank, an error will be logged and the rate will not be adjusted.
- A test for checking the probability in exponential mode
- A test for checking the probability in multiplier mode
- An unrelated test for checking that time_last_active is set to -1 for input value of time_last_active_drug_use less than -1.

ddbaptiste reviewed (changes requested):
I might've said something different yesterday (sorry), but I think we should call the new config key `linking.scaling_type` so that it more accurately captures what happens when not set to `exponential`.

I also think there needs to be a check for whether the value of the string is empty, so that the user can be notified that the functionality they might be expecting is incorrect.

Then, there's the one remaining change, commented inline below, to address.

MJC598 commented on src/event/hcv/linking.cpp:
```diff
@@ -50,6 +50,7 @@ void LinkingImpl::LoadData(datamanagement::ModelData &model_data) {
         "linking.recent_screen_multiplier", model_data));
     SetRecentScreenCutoff(
         utils::GetIntFromConfig("linking.recent_screen_cutoff", model_data));
+    SetDecayType(utils::GetStringFromConfig("linking.decay_type", model_data));
```
I'm a bit confused, I get their usage is mutually exclusive, but should we really be enforcing that in the config?

MJC598 replied:
Another reason I don't like these values in the config. Setting isn't an issue since its already garbage but "optional" values are a different animal. :disappointed:

ddbaptiste commented on src/event/internals/linking_internals.hpp:
```diff
@@ -55,14 +55,17 @@ class LinkingBase : public EventBase {
-        // apply the multiplier to recently screened persons
-        if (recently_screened && (prob < 1)) {
-            prob = ApplyMultiplier(prob, _recent_screen_multiplier);
+        if (prob < 1) {
+            bool recently_screened = (time_diff_ls <= _recent_screen_cutoff);
+            if (GetDecayType() == "exponential") {
+                prob = ApplyExpDecay(prob, time_diff_ls);
+            } else if (recently_screened && GetDecayType() == "multiplier") {
+                prob = ApplyMultiplier(prob, _recent_screen_multiplier);
+            }
```
There's an asymptote for natural log when the prbability is 1 (`ln(1 - p) = ln(0)`) where we can't do either operation.

aclark29 commented:
- Changed decay_type to scaling_type.
- Sets scaling_type's value and only sets recent multiplier and recent cutoff if scaling_type is "multiplier".
- If the string returned by `GetStringFromConfig` is empty when reading the value for scaling_type, a warning is logged saying the linking behavior might be unexpected.
- If scaling_type is not in the sim.conf file, and error will also be reported saying there was an issue extracting linking.scaling_type from config file.

- Changed time_last_active to take on whatever value for time_last_active_drug_use is passed to it, including values less than -1. (I got rid of the SetBehaviorTimeLastActive function and replaced it with a single line, but I put it in front of SetBehaviorDetails, which changes time_last_active to 0 if the person is an active drug user.)
- Changed the accompanying test to reflect that change.
